### ISSUE TITLE
Pylance Language Server gets "stuck" with bad cache, won't reparse changed file fully, requires language server restart

### ISSUE THREAD

**Issue by LightCC** [microsoft/pylance-release#2072]
#### VS Code Help Info:
Version: 1.62.1 (system setup)
Commit: f4af3cbf5a99787542e2a30fe1fd37cd644cc31f
Date: 2021-11-05T10:57:55.946Z
Electron: 13.5.2
Chrome: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Windows_NT x64 10.0.18363

-   Language Server version: v2021.11.1 (from extension window)
-   OS and version: See above
-   Python version (& distribution if applicable, e.g. Anaconda): 3.7.9

## Expected behaviour

Pylance to continually, and correctly, update its analysis as files are edited.

## Actual behaviour

Pylance linting regularly get confused or "stuck", and is partially giving results of both old and new file changes. It appears to get "stuck" with bad/partial parsing feedback and will not reanalyze a file upon save or even close and reopen of a file (seems to have it cached).

The command: `Python: Restart Language Server` resets everything and it works for a little while before having the same issue.

I have not been able to play with settings, extensions, or attempt detailed logging to see if the issue can be narrowed down. I have a non-trivial codebase this is failing on and since I have a workaround of regularly restarting the language server, have not been able to attempt to create a minimal reproducible code sample.

I do not believe I saw this behavior until a recent VS Code update (not sure if it was this most recent one for the version I am currently running, or maybe one prior to this),, and was using Pylance a decent amount, although sporadically, over the past few months. I have a fairly new employee who recently got all setup with the same setup as myself (in general, with small differences), and I saw this issue on his PC a few days ago, so it does not appear to be something specific to my PC. This occurs on more than one of my team's Python projects/repos for myself as well.

I generally have multiple VS Code instances open to multiple projects.

## Logs

I added the trace logging property to the `.vscode/settings.json` in the root of my git repo for one of the projects I'm experiencing this on, but I get this error:
![image](https://user-images.githubusercontent.com/19731713/141720329-7c3d8fcf-3cf7-47fd-92c4-7b9b92b04604.png)

However, I was able to add it to the workspace file.

I don't see much going on in the log beyond stuff like this:
```
Background analysis message: resumeAnalysis
[BG(1)] analyzing: d:\code\mca_tools\tool_apl_idl\conftest.py ...
[BG(1)]   parsing: d:\code\mca_tools\tool_apl_idl\conftest.py [fs read 0ms] (0ms)
[BG(1)]   binding: d:\code\mca_tools\tool_apl_idl\conftest.py (0ms)
[BG(1)]   checking: d:\code\mca_tools\tool_apl_idl\conftest.py ...
[BG(1)]     parsing: d:\code\mca_tools\tool_apl_idl\.venv\Lib\site-packages\_pytest\config\__init__.py [fs read 0ms] (16ms)
[BG(1)]     binding: d:\code\mca_tools\tool_apl_idl\.venv\Lib\site-packages\_pytest\config\__init__.py (6ms)
[BG(1)]     parsing: d:\code\mca_tools\tool_apl_idl\.venv\Lib\site-packages\pluggy\__init__.py [fs read 1ms] (1ms)
[BG(1)]     binding: d:\code\mca_tools\tool_apl_idl\.venv\Lib\site-packages\pluggy\__init__.py (0ms)
[BG(1)]     parsing: d:\code\mca_tools\tool_apl_idl\.venv\Lib\site-packages\pluggy\_hooks.py [fs read 1ms] (3ms)
[BG(1)]     binding: d:\code\mca_tools\tool_apl_idl\.venv\Lib\site-packages\pluggy\_hooks.py (1ms)
[BG(1)]   checking: d:\code\mca_tools\tool_apl_idl\conftest.py (31ms)
[BG(1)] analyzing: d:\code\mca_tools\tool_apl_idl\conftest.py (31ms)
[BG(1)] analyzing: d:\code\mca_tools\tool_apl_idl\run_me.py ...
[BG(1)]   parsing: d:\code\mca_tools\tool_apl_idl\run_me.py [fs read 1ms] (2ms)
[BG(1)]   binding: d:\code\mca_tools\tool_apl_idl\run_me.py (0ms)
[BG(1)]   checking: d:\code\mca_tools\tool_apl_idl\run_me.py ...
[BG(1)]     parsing: c:\Users\Mike.Leiter\.vscode\extensions\ms-python.vscode-pylance-2021.11.1\dist\typeshed-fallback\stdlib\this.pyi [fs read 1ms] (1ms)
[BG(1)]     binding: c:\Users\Mike.Leiter\.vscode\extensions\ms-python.vscode-pylance-2021.11.1\dist\typeshed-fallback\stdlib\this.pyi (0ms)
[BG(1)]     parsing: c:\Users\Mike.Leiter\.vscode\extensions\ms-python.vscode-pylance-2021.11.1\dist\typeshed-fallback\stdlib\antigravity.pyi [fs read 0ms] (0ms)
[BG(1)]     binding: c:\Users\Mike.Leiter\.vscode\extensions\ms-python.vscode-pylance-2021.11.1\dist\typeshed-fallback\stdlib\antigravity.pyi (1ms)
[BG(1)]   checking: d:\code\mca_tools\tool_apl_idl\run_me.py (3ms)
[BG(1)] analyzing: d:\code\mca_tools\tool_apl_idl\run_me.py (5ms)
[BG(1)] analyzing: d:\code\mca_tools\tool_apl_idl\src\idl\TLog.py ...
[BG(1)]   checking: d:\code\mca_tools\tool_apl_idl\src\idl\TLog.py ...
[BG(1)]     parsing: c:\Users\Mike.Leiter\.vscode\extensions\ms-python.vscode-pylance-2021.11.1\dist\typeshed-fallback\stdlib\datetime.pyi [fs read 0ms] (3ms)
[BG(1)]     binding: c:\Users\Mike.Leiter\.vscode\extensions\ms-python.vscode-pylance-2021.11.1\dist\typeshed-fallback\stdlib\datetime.pyi (2ms)
[BG(1)]     parsing: c:\Users\Mike.Leiter\.vscode\extensions\ms-python.vscode-pylance-2021.11.1\dist\typeshed-fallback\stdlib\shutil.pyi [fs read 0ms] (2ms)
[BG(1)]     binding: c:\Users\Mike.Leiter\.vscode\extensions\ms-python.vscode-pylance-2021.11.1\dist\typeshed-fallback\stdlib\shutil.pyi (0ms)
[BG(1)]   checking: d:\code\mca_tools\tool_apl_idl\src\idl\TLog.py (18ms)
[BG(1)] analyzing: d:\code\mca_tools\tool_apl_idl\src\idl\TLog.py (18ms)
[BG(1)] analyzing: d:\code\mca_tools\tool_apl_idl\src\idl\__init__.py ...
[BG(1)]   parsing: d:\code\mca_tools\tool_apl_idl\src\idl\__init__.py [fs read 1ms] (1ms)
[BG(1)]   binding: d:\code\mca_tools\tool_apl_idl\src\idl\__init__.py (0ms)
[BG(1)]   checking: d:\code\mca_tools\tool_apl_idl\src\idl\__init__.py (0ms)
[BG(1)] analyzing: d:\code\mca_tools\tool_apl_idl\src\idl\__init__.py (1ms)
[BG(1)] analyzing: d:\code\mca_tools\tool_apl_idl\src\kvconfig\__init__.py ...
[BG(1)]   checking: d:\code\mca_tools\tool_apl_idl\src\kvconfig\__init__.py (0ms)
[BG(1)] analyzing: d:\code\mca_tools\tool_apl_idl\src\kvconfig\__init__.py (0ms)
[BG(1)] analyzing: d:\code\mca_tools\tool_apl_idl\src\kvconfig\tests\__init__.py ...
[BG(1)]   parsing: d:\code\mca_tools\tool_apl_idl\src\kvconfig\tests\__init__.py [fs read 0ms] (0ms)
[BG(1)]   binding: d:\code\mca_tools\tool_apl_idl\src\kvconfig\tests\__init__.py (0ms)
[BG(1)]   checking: d:\code\mca_tools\tool_apl_idl\src\kvconfig\tests\__init__.py (0ms)
[BG(1)] analyzing: d:\code\mca_tools\tool_apl_idl\src\kvconfig\tests\__init__.py (0ms)
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
[FG] parsing: d:\code\mca_tools\tool_apl_idl\src\kvconfig\KvConfig.py (19ms)
[FG] parsing: c:\Users\Mike.Leiter\.vscode\extensions\ms-python.vscode-pylance-2021.11.1\dist\typeshed-fallback\stdlib\builtins.pyi [fs read 1ms] (16ms)
[FG] binding: c:\Users\Mike.Leiter\.vscode\extensions\ms-python.vscode-pylance-2021.11.1\dist\typeshed-fallback\stdlib\builtins.pyi (12ms)
[FG] binding: d:\code\mca_tools\tool_apl_idl\src\kvconfig\KvConfig.py (0ms)
```

## Code Snippet / Additional information

Some things that may be relevant:

#### Example 1 - Intellisense impact

Note the weird mis-coloration that occurs in this code screenshot after the issue presents, below. This includes words switching between different colors in the middle of words multiple times, with multiple different color types. It's possible we are seeing the coloration shifted by a few rows (the colors are right for a different row), given that the ghost errors/warnings appear to also be shifted (see example 2):

![image](https://user-images.githubusercontent.com/19731713/141718961-1645368f-1ffe-4f03-bd96-50f5c8efec7a.png)

#### Example 2 - Ghost Linting errors

I added a new (fake) function - this error especially seems to generate linting errors when a new function or method is added to code after it happens. All the syntax is correct from an indentation perspective, etc., and all this code was linting with no errors right before this:

![image](https://user-images.githubusercontent.com/19731713/141719146-ef5adabd-39fe-4faf-ba19-73d983e99574.png)

And the corresponding errors in the "Problems" window:
![image](https://user-images.githubusercontent.com/19731713/141719319-62e5897f-fe18-41e5-9b49-7091f1222772.png)

In the above, the final warning ("csv_encoding" is not defined), is actually legitimate, but the yellow squiggly is not in the right place!  It's reporting position [48, 34], but the actual error is on line 46 ([46, 34]).

After restarting the Language Server, the mistaken errors go away, and the legit warning has moved to the correct location, and all code coloration/formatting issues are resolved.

![image](https://user-images.githubusercontent.com/19731713/141720096-667042c0-1a4f-448f-ab8d-8ad7c168ec69.png)

#### Example Requirements.txt
Note that I have installed the local project with pip in editable mode (i.e. `py -m pip install -U -e .` from the root of the repo), and am using a `src` folder, with 2 package underneath it. I have pytest running with auto-discovery, but defaulting to no coverage, and run with coverage manually from the command line.

There don't appear to be any issues with the code executing or debugging. Everything seems to be okay other than the Pylance linting and the code language formatting as shown in the examples above.

```py
-e git+ssh://git@xxx/xxx.git@xxxHASHxxx#egg=xxx
astroid==2.8.5
atomicwrites==1.4.0
attrs==21.2.0
colorama==0.4.4
coverage==6.1.2
importlib-metadata==4.8.2
iniconfig==1.1.1
isort==5.10.1
lazy-object-proxy==1.6.0
mccabe==0.6.1
packaging==21.2
platformdirs==2.4.0
pluggy==1.0.0
py==1.11.0
pylint==2.11.1
pyparsing==2.4.7
pytest==6.2.5
pytest-cov==3.0.0
pytest-runner==5.3.1
toml==0.10.2
tomli==1.2.2
typed-ast==1.5.0
typing-extensions==3.10.0.2
wrapt==1.13.3
yapf==0.31.0
zipp==3.6.0
```

```python
XXX
```

**Comment by judej** [comment]
@LightCC, thanks for the report. This is likely a bug in our parser. Would you be able to send us the code file so that we can reproduce it here? Thanks

**Comment by erictraut** [comment]
This doesn't sound like a parser bug. It's too non-deterministic (works some times, not others). Smells more like a data corruption between the client and the language server. It's odd that this is the first report that we've heard of it though.

@LightCC, under what circumstance does it occur? When first opening your project? After editing the code in a file? After pasting code? After performing a code action or "replace all" operation? If you can discern any pattern, that would be a helpful clue.

Also, if you close and reopen an affected file, does the problem go away? If no, does modifying the file (e.g. adding and then deleting a single character) make the problem go away?

What file encoding does VS Code think the file is using? Is it "UTF-8" or something less common? (The encoding is displayed in the lower right part of the VS Code window.)

**Comment by heejaechang** [comment]
I think best way to figure out is if we can get from users the text file that shows the issue so that we can get live repro. we time to time got bug report on text and parse tree got out of sync. so far we fixed some but there might be more.

**Comment by LightCC** [comment]
Hey guys, can't give up the code files in complete format as they are work files. I'm happy to try anything or run logs, etc. or try something. I did this partly in case others were seeing the same thing or if there was anything I could do with logs or to capture it, and partly hoping there was a workaround, etc.

I do have word wrap on and I've seen some bugs with that in the past - just an idea, since the error seems to misplace where the code is, rather than operating on old code, it might just be shifted due to word wrap. I remember something about changes to word wrap a few versions ago.

@erictraut Mostly utf-8 encoding, though some of the CSVs are utf-8-sig. I constantly have a ton of files open with multiple workspaces, have my caches turned up a bit and Code was using 5-6 Gb of memory last I checked, but it's all pretty bland, as far as I can tell.

I was struggling a few weeks back with importing as I have multiple packages in this project, but got that all straightened out and have the project properly installed as a Python package now (`py -m pip install -e .' in the root) using setup.py, installed to a venv. It's a mix of package, relative and `__init__.py` imports.

I haven't been able to discern any patterns, beyond what I described above. It happens when it happens, but seems to be kicked off by editing a file, especially in a way that adds new lines, functions and indentation blocks. It does NOT appear to ever reset itself, and I tried a lot of things - i.e. you can't just ctrl-z or manually reverse the edit to get it back, so it seems to be going into a bad state when this happens. Until I searched and found out about the Language Server Restart command I was tearing my hair out and tried quite a few ways of potentially forcing it (close/reopen files, close/reopen vs code).

Today on primarily some other files in the same repo compared to yesterday it's mostly gone away, but last night it seemed like it was showing up very regularly (enough that I mapped Restart Language Server to a key combo).

**Comment by heejaechang** [comment]
@LightCC any update?

**Comment by heejaechang** [comment]
please re-open if this is still an issue.

**Comment by zcattacz** [comment]
I also got this Ghost Linting problem. after the pylance extension update and reload, initially the linter seems work fine and after a dozen of seconds, I start to get non-sense lint errors, e.g. indentation error  for the first few lines of `import`. It seems that these errors are for a complete different file. I tried reload the window from F1 drop down menu, but it didn't help. The linter didn't regains sanity, until I choose restart language server like the OP suggested in this issue.

**Comment by sleeplessghost** [comment]
FYI @zcattacz I experienced the same issue, appears to be introduced in version 2022.11.30, fixed in (Pre-Release) 2022.11.41 (at least it is working for me)

**Comment by zcattacz** [comment]
@sleeplessghost  Thanks, good to know. But since it has been happening for quite some time and I found a quicker fix than reload, It no longer bothers me that much : rename the file to arbitrary name wait 1 second, and rename it back, the linter will immediately regain sanity.
